## The cost of milk to make cheese
<p>The Coronavirus outbreak has caused major economic disruption. The pandemic's impact on the tourism, travel, and hospitality industries has featured frequently in 2020 news reports, but the economic impact has not been isolated to these industries alone. </p>
<p>The UK dairy industry is one such affected industry and has had to work together to address current market challenges and avoid milk waste. The industry can adapt to the temporary decrease in demand by identifying opportunities for processing the milk into storable products such as butter, cheese, and skimmed milk powder. </p>
<p>The UK Department of Environment, Food, and Rural Affairs has contracted your company to perform a detailed analysis of the UK dairy industry to help them understand the industry's value. </p>
<p>Your team has divided the analysis project into several tasks. One of the tasks is to determine the value of the cheddar cheese industry. The revenue from cheese is much more than, for example, whole milk, but a kilogram of cheese requires almost ten liters of raw milk to make. Your first task is to determine the annual cost of milk needed to produce cheddar cheese over the past five years. </p>
<p>You have been provided with three datasets. <em><strong>Please note that 1000 kg = 1 metric tonne.</strong></em></p>
<div style="background-color: #efebe4; color: #05192d; text-align:left; vertical-align: middle; padding: 15px 25px 15px 25px; line-height: 1.6;">
    <div style="font-size:16px"><b>datasets/milk_prices.csv - The monthly prices, volume, protein percentage and butterfat percentage of milk </b>
    </div>
Source: <a href="https://environment.data.gov.uk/linked-data/cube/explore?uri=http%3A%2F%2Fenvironment.data.gov.uk%2Flinked-data%2Fcatalog%2Fdatasets%2Fentry%2Fmilk-prices-and-composition-of-milk-annual-statistics&filters-drawer=closed">Defra Data Services Platform</a>
<ul>
    <li><b>Time: </b>The month and year when the value was recorded.</li>
    <li><b>Measure Type: </b>The type of measure recorded.</li>
    <li><b>Price:</b> The price of milk (in Pounds per liter).</li>
    <li><b>Volume:</b> The amount of milk produced (in million liters).</li>
    <li><b>Protein:</b> The protein content of the milk (percentage).</li>
    <li><b>Butterfat:</b> The butterfat content of the milk (percentage).</li>
    <li><b>Unit of Measure:</b> The units specific to the measure type.</li>
</ul>
</div>
<div style="background-color: #efebe4; color: #05192d; text-align:left; vertical-align: middle; padding: 15px 25px 15px 25px; line-height: 1.6; margin-top: 17px;">
    <div style="font-size:16px"><b>datasets/Milk_products_production.csv - The monthly production figures of milk, cream and cheddar cheese</b>
    </div>
Source: <a href="https://www.gov.uk/government/statistics/milk-utilisation-by-dairies-in-england-and-wales">GOV.UK</a>
    <ul>
        <li><b>Unnamed: 0: </b>The year and month when the value was recorded.</li>
        <li><b>Liquid Milk Production: </b>The total amount of milk produced (in million liters).</li>
        <li><b>Cream Production: </b>The total amount of cream produced (in million liters).</li>
        <li><b>Cheddar Cheese Production:</b>The total amount of cheddar cheese produced (in thousand tonnes).</li>
    </ul>
</div>
<div style="background-color: #efebe4; color: #05192d; text-align:left; vertical-align: middle; padding: 15px 25px 15px 25px; line-height: 1.6; margin-top: 17px;">    
    <div style="font-size:16px"><b>datasets/conversion_factors.xls - Liters of milk used to make one kilogram of product</b>
    </div>
Source: <a href="https://www.gov.uk/government/statistics/milk-utilisation-by-dairies-in-england-and-wales">GOV.UK</a>
    <ul>
        <li><b>Product: </b>The type of product.</li>
        <li><b>Conversion factor (litres/kg):</b> Liters of milk used to make one kilogram of product (in liters/kg).</li>
    </ul>
</div>

## Imports

In [162]:
import pandas as pd

In [163]:
# Loading data
milk_prices = pd.read_csv("datasets/milk_prices.csv")
milk_products_production = pd.read_csv('datasets/Milk_products_production.csv')
conversion_factors = pd.read_excel('datasets/conversion_factors.xls')

## Milk Prices DataFrame Analysis

### First look

In [164]:
milk_prices

Time Measure type  Price  Volume  Protein  Butterfat   Unit of Measure
0     1986-10        Price  0.170     NaN      NaN        NaN  Pounds per litre
1     2000-04       Volume    NaN  1177.0      NaN        NaN    Million litres
2     1994-12      Protein    NaN     NaN     3.21        NaN        Percentage
3     1997-10        Price  0.208     NaN      NaN        NaN  Pounds per litre
4     2002-10      Protein    NaN     NaN     3.41        NaN        Percentage
...       ...          ...    ...     ...      ...        ...               ...
1552  2004-07       Volume    NaN  1203.0      NaN        NaN    Million litres
1553  2008-11        Price  0.271     NaN      NaN        NaN  Pounds per litre
1554  2019-11        Price  0.299     NaN      NaN        NaN  Pounds per litre
1555  2009-07      Protein    NaN     NaN     3.22        NaN        Percentage
1556  2016-08       Volume    NaN  1126.0      NaN        NaN    Million litres

[1557 rows x 7 columns]

In [165]:
milk_prices.describe()

Price    Volume  Protein  Butterfat
count  611.000   313.000  313.000    320.000
mean     0.184  1153.058    3.292      4.044
std      0.075    88.408    0.065      0.113
min      0.029   973.000    3.120      3.760
25%      0.138  1078.000    3.240      3.960
50%      0.183  1148.000    3.290      4.060
75%      0.243  1206.000    3.330      4.130
max      0.345  1395.000    3.470      4.310

In [166]:
milk_prices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1557 entries, 0 to 1556
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Time             1557 non-null   object 
 1   Measure type     1557 non-null   object 
 2   Price            611 non-null    float64
 3   Volume           313 non-null    float64
 4   Protein          313 non-null    float64
 5   Butterfat        320 non-null    float64
 6   Unit of Measure  1557 non-null   object 
dtypes: float64(4), object(3)
memory usage: 85.3+ KB


#### What to keep in mind?
* Missing values - all the columns except Time and Measure type.
* Std and mean - differ
* There are some not necessary columns/or columns for fe (measure.., protein, butterfat)
* Time column - first step need to be dealt with. 

It seems like in the milk_prices df they are also entries from the year out of our desired range. We don't need any of these. What we'll do?
1. First we'll make a new column based only on the year (we don't need a month)
2. Second we'll select only rows which the year is betweem 2015-2020. 

### Feature engineering

In [167]:
# Creating temp datetime column
milk_prices['date'] = pd.to_datetime(milk_prices['Time'])

In [168]:
# Dropping original Time column
milk_prices.drop('Time', axis=1, inplace=True)

In [169]:
# Checking the result
milk_prices.head()

Measure type  Price  Volume  Protein  Butterfat   Unit of Measure       date
0        Price  0.170     NaN      NaN        NaN  Pounds per litre 1986-10-01
1       Volume    NaN  1177.0      NaN        NaN    Million litres 2000-04-01
2      Protein    NaN     NaN     3.21        NaN        Percentage 1994-12-01
3        Price  0.208     NaN      NaN        NaN  Pounds per litre 1997-10-01
4      Protein    NaN     NaN     3.41        NaN        Percentage 2002-10-01

In [170]:
# Creating Year datetime column
milk_prices['Year'] = pd.DatetimeIndex(milk_prices['date']).year
milk_prices

Measure type  Price  Volume  Protein  Butterfat   Unit of Measure       date  Year
0           Price  0.170     NaN      NaN        NaN  Pounds per litre 1986-10-01  1986
1          Volume    NaN  1177.0      NaN        NaN    Million litres 2000-04-01  2000
2         Protein    NaN     NaN     3.21        NaN        Percentage 1994-12-01  1994
3           Price  0.208     NaN      NaN        NaN  Pounds per litre 1997-10-01  1997
4         Protein    NaN     NaN     3.41        NaN        Percentage 2002-10-01  2002
...           ...    ...     ...      ...        ...               ...        ...   ...
1552       Volume    NaN  1203.0      NaN        NaN    Million litres 2004-07-01  2004
1553        Price  0.271     NaN      NaN        NaN  Pounds per litre 2008-11-01  2008
1554        Price  0.299     NaN      NaN        NaN  Pounds per litre 2019-11-01  2019
1555      Protein    NaN     NaN     3.22        NaN        Percentage 2009-07-01  2009
1556       Volume    NaN  1126.0      NaN        NaN    Million litres 2016-08-01  2016

[1557 rows x 8 columns]

In [171]:
milk_prices['Month']= pd.DatetimeIndex(milk_prices['date']).month
milk_prices

Measure type  Price  Volume  Protein  Butterfat   Unit of Measure       date  Year  Month
0           Price  0.170     NaN      NaN        NaN  Pounds per litre 1986-10-01  1986     10
1          Volume    NaN  1177.0      NaN        NaN    Million litres 2000-04-01  2000      4
2         Protein    NaN     NaN     3.21        NaN        Percentage 1994-12-01  1994     12
3           Price  0.208     NaN      NaN        NaN  Pounds per litre 1997-10-01  1997     10
4         Protein    NaN     NaN     3.41        NaN        Percentage 2002-10-01  2002     10
...           ...    ...     ...      ...        ...               ...        ...   ...    ...
1552       Volume    NaN  1203.0      NaN        NaN    Million litres 2004-07-01  2004      7
1553        Price  0.271     NaN      NaN        NaN  Pounds per litre 2008-11-01  2008     11
1554        Price  0.299     NaN      NaN        NaN  Pounds per litre 2019-11-01  2019     11
1555      Protein    NaN     NaN     3.22        NaN        Percentage 2009-07-01  2009      7
1556       Volume    NaN  1126.0      NaN        NaN    Million litres 2016-08-01  2016      8

[1557 rows x 9 columns]

In [172]:
# Dropping temp date datetime column
milk_prices.drop('date', inplace=True, axis=1)

In [173]:
# Checking the result
milk_prices.head()

Measure type  Price  Volume  Protein  Butterfat   Unit of Measure  Year  Month
0        Price  0.170     NaN      NaN        NaN  Pounds per litre  1986     10
1       Volume    NaN  1177.0      NaN        NaN    Million litres  2000      4
2      Protein    NaN     NaN     3.21        NaN        Percentage  1994     12
3        Price  0.208     NaN      NaN        NaN  Pounds per litre  1997     10
4      Protein    NaN     NaN     3.41        NaN        Percentage  2002     10

In [174]:
# Filtering rows based on desired years 2015-2020
milk_prices_FILTERED = milk_prices.loc[(milk_prices['Year'] >= 2015) 
                                       & (milk_prices['Year'] <= 2020)]

In [175]:
# Checking the results
milk_prices_FILTERED

Measure type  Price  Volume  Protein  Butterfat   Unit of Measure  Year  Month
19         Volume    NaN  1211.0      NaN        NaN    Million litres  2015      8
25         Volume    NaN  1078.0      NaN        NaN    Million litres  2016      9
26          Price  0.294     NaN      NaN        NaN  Pounds per litre  2019     10
31        Protein    NaN     NaN     3.36        NaN        Percentage  2019      5
34        Protein    NaN     NaN     3.27        NaN        Percentage  2018      4
...           ...    ...     ...      ...        ...               ...   ...    ...
1510      Protein    NaN     NaN     3.29        NaN        Percentage  2016      3
1519      Protein    NaN     NaN     3.27        NaN        Percentage  2015      2
1550       Volume    NaN  1267.0      NaN        NaN    Million litres  2015      7
1554        Price  0.299     NaN      NaN        NaN  Pounds per litre  2019     11
1556       Volume    NaN  1126.0      NaN        NaN    Million litres  2016      8

[284 rows x 8 columns]

In [176]:
milk_prices_FILTERED.Year.unique()

array([2015, 2016, 2019, 2018, 2020, 2017])

Seems all good.

Year is filtered and now we are dealing with only 284 rows. 

---


In [177]:
# Checking non null Values in Price columns
milk_prices_FILTERED[milk_prices_FILTERED.Price.notnull()].head(50)

Measure type  Price  Volume  Protein  Butterfat   Unit of Measure  Year  Month
26          Price  0.294     NaN      NaN        NaN  Pounds per litre  2019     10
451         Price  0.290     NaN      NaN        NaN  Pounds per litre  2020      9
481         Price  0.281     NaN      NaN        NaN  Pounds per litre  2020      8
506         Price  0.277     NaN      NaN        NaN  Pounds per litre  2020      7
515         Price  0.271     NaN      NaN        NaN  Pounds per litre  2020      6
542         Price  0.267     NaN      NaN        NaN  Pounds per litre  2020      5
549         Price  0.274     NaN      NaN        NaN  Pounds per litre  2020      4
569         Price  0.238     NaN      NaN        NaN  Pounds per litre  2015      9
584         Price  0.286     NaN      NaN        NaN  Pounds per litre  2020      3
603         Price  0.287     NaN      NaN        NaN  Pounds per litre  2020      2
615         Price  0.227     NaN      NaN        NaN  Pounds per litre  2016      9
618         Price  0.234     NaN      NaN        NaN  Pounds per litre  2015      8
624         Price  0.304     NaN      NaN        NaN  Pounds per litre  2017      9
628         Price  0.215     NaN      NaN        NaN  Pounds per litre  2016      8
630         Price  0.236     NaN      NaN        NaN  Pounds per litre  2015      7
659         Price  0.288     NaN      NaN        NaN  Pounds per litre  2020      1
686         Price  0.306     NaN      NaN        NaN  Pounds per litre  2020     11
695         Price  0.306     NaN      NaN        NaN  Pounds per litre  2018      9
699         Price  0.291     NaN      NaN        NaN  Pounds per litre  2017      8
703         Price  0.208     NaN      NaN        NaN  Pounds per litre  2016      7
706         Price  0.238     NaN      NaN        NaN  Pounds per litre  2015      6
731         Price  0.199     NaN      NaN        NaN  Pounds per litre  2016      6
738         Price  0.241     NaN      NaN        NaN  Pounds per litre  2015      5
779         Price  0.300     NaN      NaN        NaN  Pounds per litre  2020     10
784         Price  0.292     NaN      NaN        NaN  Pounds per litre  2019      9
790         Price  0.297     NaN      NaN        NaN  Pounds per litre  2018      8
795         Price  0.279     NaN      NaN        NaN  Pounds per litre  2017      7
810         Price  0.286     NaN      NaN        NaN  Pounds per litre  2019      8
818         Price  0.286     NaN      NaN        NaN  Pounds per litre  2018      7
825         Price  0.268     NaN      NaN        NaN  Pounds per litre  2017      6
834         Price  0.204     NaN      NaN        NaN  Pounds per litre  2016      5
840         Price  0.247     NaN      NaN        NaN  Pounds per litre  2015      4
879         Price  0.251     NaN      NaN        NaN  Pounds per litre  2015      3
926         Price  0.284     NaN      NaN        NaN  Pounds per litre  2019      7
933         Price  0.272     NaN      NaN        NaN  Pounds per litre  2018      6
944         Price  0.268     NaN      NaN        NaN  Pounds per litre  2017      5
954         Price  0.216     NaN      NaN        NaN  Pounds per litre  2016      4
972         Price  0.280     NaN      NaN        NaN  Pounds per litre  2019      6
984         Price  0.268     NaN      NaN        NaN  Pounds per litre  2018      5
994         Price  0.269     NaN      NaN        NaN  Pounds per litre  2017      4
1002        Price  0.224     NaN      NaN        NaN  Pounds per litre  2016      3
1010        Price  0.261     NaN      NaN        NaN  Pounds per litre  2015      2
1044        Price  0.265     NaN      NaN        NaN  Pounds per litre  2015      1
1098        Price  0.278     NaN      NaN        NaN  Pounds per litre  2019      5
1109        Price  0.274     NaN      NaN        NaN  Pounds per litre  2018      4
1118        Price  0.275     NaN      NaN        NaN  Pounds per litre  2017      3
1125        Price  0.230     NaN      NaN        NaN  Pou

Okay. So it seems, like I thought. Price value is present only in the rows with Price measurement type. So the columns Proetin and Butterfat are unnecessary. An with this also the measure type column. 

We'll drop unnecessary columns Protein and Butterfat, Mesaure Type, Volume. 

In [178]:
milk_prices_FILTERED.drop('Protein', axis=1, inplace=True)
milk_prices_FILTERED.drop('Butterfat', axis=1, inplace=True)
milk_prices_FILTERED.drop('Measure type', axis=1, inplace=True)
milk_prices_FILTERED.drop('Volume', axis=1, inplace=True)

In [179]:
milk_prices_FILTERED = milk_prices_FILTERED[milk_prices_FILTERED.Price.notnull()]

In [180]:
milk_prices_FILTERED

Price   Unit of Measure  Year  Month
26    0.294  Pounds per litre  2019     10
451   0.290  Pounds per litre  2020      9
481   0.281  Pounds per litre  2020      8
506   0.277  Pounds per litre  2020      7
515   0.271  Pounds per litre  2020      6
...     ...               ...   ...    ...
1453  0.296  Pounds per litre  2019      1
1458  0.316  Pounds per litre  2018     11
1461  0.318  Pounds per litre  2017     10
1490  0.315  Pounds per litre  2018     10
1554  0.299  Pounds per litre  2019     11

[71 rows x 4 columns]

In [181]:
milk_prices_FILTERED.describe()

Price      Year   Month
count  71.000    71.000  71.000
mean    0.271  2017.465   6.423
std     0.030     1.706   3.438
min     0.199  2015.000   1.000
25%     0.246  2016.000   3.500
50%     0.275  2017.000   6.000
75%     0.293  2019.000   9.000
max     0.320  2020.000  12.000

In [182]:
milk_prices_FILTERED.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 71 entries, 26 to 1554
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Price            71 non-null     float64
 1   Unit of Measure  71 non-null     object 
 2   Year             71 non-null     int64  
 3   Month            71 non-null     int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 2.8+ KB


Better. 

Now it seems like, we should only select rows with Measure Type equal to Price. I wonder about Volume though. Can we do something with other data also? I don't think so, because. Protein, Butterfat have no info about 

In [183]:
milk_prices_FILTERED['Unit of Measure'].unique()

array(['Pounds per litre'], dtype=object)

It seems, like with out current selection there is no other Unit of measure as Pounds per litre. 

We'll keep that in mind, that our price is here measured in Pounds per litre, so maybe renaming Price columns should be in order?

Let's do it. 

In [184]:
# Creating Price columns with different name (for our convenience)
milk_prices_FILTERED['Price in Pounds per Litre'] = milk_prices_FILTERED['Price']
milk_prices_FILTERED

Price   Unit of Measure  Year  Month  Price in Pounds per Litre
26    0.294  Pounds per litre  2019     10                      0.294
451   0.290  Pounds per litre  2020      9                      0.290
481   0.281  Pounds per litre  2020      8                      0.281
506   0.277  Pounds per litre  2020      7                      0.277
515   0.271  Pounds per litre  2020      6                      0.271
...     ...               ...   ...    ...                        ...
1453  0.296  Pounds per litre  2019      1                      0.296
1458  0.316  Pounds per litre  2018     11                      0.316
1461  0.318  Pounds per litre  2017     10                      0.318
1490  0.315  Pounds per litre  2018     10                      0.315
1554  0.299  Pounds per litre  2019     11                      0.299

[71 rows x 5 columns]

In [185]:
# Dropping unnecessary columns
milk_prices_FILTERED.drop('Price', axis=1, inplace=True)
milk_prices_FILTERED.drop('Unit of Measure', axis=1, inplace=True)

In [186]:
# Final result
milk_prices_FILTERED.head(50)

Year  Month  Price in Pounds per Litre
26    2019     10                      0.294
451   2020      9                      0.290
481   2020      8                      0.281
506   2020      7                      0.277
515   2020      6                      0.271
542   2020      5                      0.267
549   2020      4                      0.274
569   2015      9                      0.238
584   2020      3                      0.286
603   2020      2                      0.287
615   2016      9                      0.227
618   2015      8                      0.234
624   2017      9                      0.304
628   2016      8                      0.215
630   2015      7                      0.236
659   2020      1                      0.288
686   2020     11                      0.306
695   2018      9                      0.306
699   2017      8                      0.291
703   2016      7                      0.208
706   2015      6                      0.238
731   2016      6                      0.199
738   2015      5                      0.241
779   2020     10                      0.300
784   2019      9                      0.292
790   2018      8                      0.297
795   2017      7                      0.279
810   2019      8                      0.286
818   2018      7                      0.286
825   2017      6                      0.268
834   2016      5                      0.204
840   2015      4                      0.247
879   2015      3                      0.251
926   2019      7                      0.284
933   2018      6                      0.272
944   2017      5                      0.268
954   2016      4                      0.216
972   2019      6                      0.280
984   2018      5                      0.268
994   2017      4                      0.269
1002  2016      3                      0.224
1010  2015      2                      0.261
1044  2015      1                      0.265
1098  2019      5                      0.278
1109  2018      4                      0.274
1118  2017      3                      0.275
1125  2016      2                      0.230
1133  2015     12                      0.240
1146  2019      4                      0.282
1154  2018      3                      0.285

## Milk Products Production Analysis

In [187]:
milk_products_production

Unnamed: 0 Liquid Milk Production Cream Production Cheddar Cheese Production
0         NaN         Million Litres   Million Litres           Thousand tonnes
1      Jan-15                  571,2             25,7                      24,2
2      Feb-15                  524,1             22,5                      21,3
3      Mar-15                  583,0             25,7                      26,2
4      Apr-15                  542,0             26,4                      29,3
..        ...                    ...              ...                       ...
68     Aug-20                  505,3             25,5                      27,4
69     Sep-20                  499,0             26,0                      26,1
70     Oct-20                  519,3             27,7                      25,6
71     Nov-20                  522,3             28,9                      25,3
72     Dec-20                  530,3             29,8                      27,3

[73 rows x 4 columns]

In this dataset we need:
1. Drop unnecessary columns Liquid Milk Production, Cream Production
    > We only need the data about Cheddar Chese Production and Year of the Production. 
2. Transform Unnamed columns into datetime column, which will store only years from 2015 to 2020. 
3. Make transfromation in the unit of measurements.
    > In the Milk Prices we have pounds per LITRE. Cheddar we have thousands of tonnes. And in the last table, we have info about how many LITERS of milk were used to make 1 KILOGRAM of cheddar. 

In [188]:
## Dropping unnecessary columns Liquid Milk Production and Cream Production
cheddar_production = milk_products_production.drop(['Liquid Milk Production', 
                                                   'Cream Production'], axis=1)

In [189]:
cheddar_production.head()

Unnamed: 0 Cheddar Cheese Production
0        NaN           Thousand tonnes
1     Jan-15                      24,2
2     Feb-15                      21,3
3     Mar-15                      26,2
4     Apr-15                      29,3

Dropping columns is done. But we can drop also the first row. It was read as a input data, but it is only storing info about the unit of measurement. Let's check though, if it is the only occurence of the unit in this column.

In [190]:
cheddar_production['Cheddar Cheese Production'].unique()

array(['Thousand tonnes', '24,2', '21,3', '26,2', '29,3', '31,0', '27,8',
       '26,0', '24,3', '22,5', '22,9', '21,9', '25,0', '23,9', '27,2',
       '28,5', '30,9', '26,8', '26,1', '21,6', '25,2', '28,9', '24,7',
       '29,8', '25,8', '28,6', '27,7', '26,4', '24,9', '26,5', '28,2',
       '27,4', '25,1', '26,9', '34,6', '31,1', '26,7', '23,8', '25,7',
       '29,5', '28,0', '27,1', '24,6', '26,3', '31,5', '34,0', '30,7',
       '25,6', '25,3', '27,3'], dtype=object)

Yes.

It is only one occurences of the unit. And it is not seemed to be inputs in others units based on the look of the values, also. 

> We can safely drop the first row. 

In [191]:
cheddar_production.drop(index=0, axis=0, inplace=True)

In [192]:
cheddar_production.head()

Unnamed: 0 Cheddar Cheese Production
1     Jan-15                      24,2
2     Feb-15                      21,3
3     Mar-15                      26,2
4     Apr-15                      29,3
5     May-15                      31,0

### Feature engineering

Next step is transforming date column. 

In [193]:
cheddar_production.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 72 entries, 1 to 72
Data columns (total 2 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Unnamed: 0                 72 non-null     object
 1   Cheddar Cheese Production  72 non-null     object
dtypes: object(2)
memory usage: 1.7+ KB


In [194]:
cheddar_production['Unnamed: 0'].unique()

array(['Jan-15', 'Feb-15', 'Mar-15', 'Apr-15', 'May-15', 'Jun-15',
       'Jul-15', 'Aug-15', 'Sep-15', 'Oct-15', 'Nov-15', 'Dec-15',
       'Jan-16', 'Feb-16', 'Mar-16', 'Apr-16', 'May-16', 'Jun-16',
       'Jul-16', 'Aug-16', 'Sep-16', 'Oct-16', 'Nov-16', 'Dec-16',
       'Jan-17', 'Feb-17', 'Mar-17', 'Apr-17', 'May-17', 'Jun-17',
       'Jul-17', 'Aug-17', 'Sep-17', 'Oct-17', 'Nov-17', 'Dec-17',
       'Jan-18', 'Feb-18', 'Mar-18', 'Apr-18', 'May-18', 'Jun-18',
       'Jul-18', 'Aug-18', 'Sep-18', 'Oct-18', 'Nov-18', 'Dec-18',
       'Jan-19', 'Feb-19', 'Mar-19', 'Apr-19', 'May-19', 'Jun-19',
       'Jul-19', 'Aug-19', 'Sep-19', 'Oct-19', 'Nov-19', 'Dec-19',
       'Jan-20', 'Feb-20', 'Mar-20', 'Apr-20', 'May-20', 'Jun-20',
       'Jul-20', 'Aug-20', 'Sep-20', 'Oct-20', 'Nov-20', 'Dec-20'],
      dtype=object)

It seems like all the date inputs are from our desired year range. So we don't need to transfrom it like in the first table. But! We'll make some adjustments, to make it nicer. 

In [195]:
# Extracting only digits (with regex) from Data column and creating new columns based on it
cheddar_production['Year'] = cheddar_production['Unnamed: 0'].str.extract('(\d+)')

In [196]:
# Checking results
cheddar_production.head()

Unnamed: 0 Cheddar Cheese Production Year
1     Jan-15                      24,2   15
2     Feb-15                      21,3   15
3     Mar-15                      26,2   15
4     Apr-15                      29,3   15
5     May-15                      31,0   15

In [197]:
# Adding prefix year column
cheddar_production['p_temp'] = '20'
cheddar_production

Unnamed: 0 Cheddar Cheese Production Year p_temp
1      Jan-15                      24,2   15     20
2      Feb-15                      21,3   15     20
3      Mar-15                      26,2   15     20
4      Apr-15                      29,3   15     20
5      May-15                      31,0   15     20
..        ...                       ...  ...    ...
68     Aug-20                      27,4   20     20
69     Sep-20                      26,1   20     20
70     Oct-20                      25,6   20     20
71     Nov-20                      25,3   20     20
72     Dec-20                      27,3   20     20

[72 rows x 4 columns]

In [198]:
# Concatenating string columns
cheddar_production['Year'] = cheddar_production.p_temp.str.cat(cheddar_production.Year)

In [199]:
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year p_temp
1      Jan-15                      24,2  2015     20
2      Feb-15                      21,3  2015     20
3      Mar-15                      26,2  2015     20
4      Apr-15                      29,3  2015     20
5      May-15                      31,0  2015     20
..        ...                       ...   ...    ...
68     Aug-20                      27,4  2020     20
69     Sep-20                      26,1  2020     20
70     Oct-20                      25,6  2020     20
71     Nov-20                      25,3  2020     20
72     Dec-20                      27,3  2020     20

[72 rows x 4 columns]

In [200]:
# Dropping old Unnamed column
#cheddar_production.drop('Unnamed: 0', axis=1, inplace=True)

In [201]:
# Dropping temp column
cheddar_production.drop('p_temp', axis=1, inplace=True)
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year
1      Jan-15                      24,2  2015
2      Feb-15                      21,3  2015
3      Mar-15                      26,2  2015
4      Apr-15                      29,3  2015
5      May-15                      31,0  2015
..        ...                       ...   ...
68     Aug-20                      27,4  2020
69     Sep-20                      26,1  2020
70     Oct-20                      25,6  2020
71     Nov-20                      25,3  2020
72     Dec-20                      27,3  2020

[72 rows x 3 columns]

In [202]:
# Transfroming into datetime column
cheddar_production['Year'] = pd.to_datetime(cheddar_production['Year']).dt.year

In [203]:
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year
1      Jan-15                      24,2  2015
2      Feb-15                      21,3  2015
3      Mar-15                      26,2  2015
4      Apr-15                      29,3  2015
5      May-15                      31,0  2015
..        ...                       ...   ...
68     Aug-20                      27,4  2020
69     Sep-20                      26,1  2020
70     Oct-20                      25,6  2020
71     Nov-20                      25,3  2020
72     Dec-20                      27,3  2020

[72 rows x 3 columns]

#### Month column creation

In [204]:
# Created new column based on Unnamed
cheddar_production['Month'] = cheddar_production['Unnamed: 0'].str[:3]
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year Month
1      Jan-15                      24,2  2015   Jan
2      Feb-15                      21,3  2015   Feb
3      Mar-15                      26,2  2015   Mar
4      Apr-15                      29,3  2015   Apr
5      May-15                      31,0  2015   May
..        ...                       ...   ...   ...
68     Aug-20                      27,4  2020   Aug
69     Sep-20                      26,1  2020   Sep
70     Oct-20                      25,6  2020   Oct
71     Nov-20                      25,3  2020   Nov
72     Dec-20                      27,3  2020   Dec

[72 rows x 4 columns]

In [205]:
# Mapping month equivalent values to the shorts
# Creating a dictionary for mapping
Month = {'Jan': '1','Feb': '2', 'Mar': '3', 'Apr': '4', 'May': '5', 'Jun': '6', 
         'Jul': '7', 'Aug': '8', 'Sep': '9', 'Oct': '10', 'Nov': '11', 'Dec': '12'}

In [206]:
cheddar_production['Month_Numeric'] = cheddar_production['Month'].map(Month)
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year Month Month_Numeric
1      Jan-15                      24,2  2015   Jan             1
2      Feb-15                      21,3  2015   Feb             2
3      Mar-15                      26,2  2015   Mar             3
4      Apr-15                      29,3  2015   Apr             4
5      May-15                      31,0  2015   May             5
..        ...                       ...   ...   ...           ...
68     Aug-20                      27,4  2020   Aug             8
69     Sep-20                      26,1  2020   Sep             9
70     Oct-20                      25,6  2020   Oct            10
71     Nov-20                      25,3  2020   Nov            11
72     Dec-20                      27,3  2020   Dec            12

[72 rows x 5 columns]

In [207]:
# Dropping Short month column, Correcting data types, Names
cheddar_production['Month'] = cheddar_production['Month_Numeric']

In [208]:
cheddar_production.drop('Month_Numeric', axis=1, inplace=True)

In [209]:
cheddar_production.head(10)

Unnamed: 0 Cheddar Cheese Production  Year Month
1      Jan-15                      24,2  2015     1
2      Feb-15                      21,3  2015     2
3      Mar-15                      26,2  2015     3
4      Apr-15                      29,3  2015     4
5      May-15                      31,0  2015     5
6      Jun-15                      27,8  2015     6
7      Jul-15                      26,0  2015     7
8      Aug-15                      24,3  2015     8
9      Sep-15                      22,5  2015     9
10     Oct-15                      22,9  2015    10

In [210]:
cheddar_production.Month.dtype

dtype('O')

In [211]:
cheddar_production['Month'] = pd.to_numeric(cheddar_production.Month)

In [212]:
cheddar_production.Month.dtype
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year  Month
1      Jan-15                      24,2  2015      1
2      Feb-15                      21,3  2015      2
3      Mar-15                      26,2  2015      3
4      Apr-15                      29,3  2015      4
5      May-15                      31,0  2015      5
..        ...                       ...   ...    ...
68     Aug-20                      27,4  2020      8
69     Sep-20                      26,1  2020      9
70     Oct-20                      25,6  2020     10
71     Nov-20                      25,3  2020     11
72     Dec-20                      27,3  2020     12

[72 rows x 4 columns]

In [213]:
# for some reason Month is changing to value one after running this code
#cheddar_production['Month'] = pd.DatetimeIndex(cheddar_production['Month']).month

In [214]:
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year  Month
1      Jan-15                      24,2  2015      1
2      Feb-15                      21,3  2015      2
3      Mar-15                      26,2  2015      3
4      Apr-15                      29,3  2015      4
5      May-15                      31,0  2015      5
..        ...                       ...   ...    ...
68     Aug-20                      27,4  2020      8
69     Sep-20                      26,1  2020      9
70     Oct-20                      25,6  2020     10
71     Nov-20                      25,3  2020     11
72     Dec-20                      27,3  2020     12

[72 rows x 4 columns]

In [215]:
cheddar_production['In thousand tonnes'] = cheddar_production['Cheddar Cheese Production']
cheddar_production

Unnamed: 0 Cheddar Cheese Production  Year  Month In thousand tonnes
1      Jan-15                      24,2  2015      1               24,2
2      Feb-15                      21,3  2015      2               21,3
3      Mar-15                      26,2  2015      3               26,2
4      Apr-15                      29,3  2015      4               29,3
5      May-15                      31,0  2015      5               31,0
..        ...                       ...   ...    ...                ...
68     Aug-20                      27,4  2020      8               27,4
69     Sep-20                      26,1  2020      9               26,1
70     Oct-20                      25,6  2020     10               25,6
71     Nov-20                      25,3  2020     11               25,3
72     Dec-20                      27,3  2020     12               27,3

[72 rows x 5 columns]

In [216]:
cheddar_production.drop('Cheddar Cheese Production', inplace=True, axis=1)
cheddar_production

Unnamed: 0  Year  Month In thousand tonnes
1      Jan-15  2015      1               24,2
2      Feb-15  2015      2               21,3
3      Mar-15  2015      3               26,2
4      Apr-15  2015      4               29,3
5      May-15  2015      5               31,0
..        ...   ...    ...                ...
68     Aug-20  2020      8               27,4
69     Sep-20  2020      9               26,1
70     Oct-20  2020     10               25,6
71     Nov-20  2020     11               25,3
72     Dec-20  2020     12               27,3

[72 rows x 4 columns]

In [217]:
# We no longer need date column
cheddar_production.drop('Unnamed: 0', inplace=True, axis=1)

Now is time to convert thousand tonnes into kilogrames. It will be a big numbers here. 

So first thing is to convert thousand tonnes into tonnes. 
    > Multiply by 1 000.
Second thing is to convert tonnes into kg. 1 tonne is 1 000 kg.
    > Multiply by 1 000.
    
Conclusion: we need to multiply the thousand tonnes by 1 000 000.

In [218]:
# Converting thousand tonnes into numerical column
# First need to replace commas with dots - python don't want to convert the columns into float type
cheddar_production['In thousand tonnes'] = cheddar_production['In thousand tonnes'].apply(lambda x: x.replace(',', '.'))
cheddar_production

Year  Month In thousand tonnes
1   2015      1               24.2
2   2015      2               21.3
3   2015      3               26.2
4   2015      4               29.3
5   2015      5               31.0
..   ...    ...                ...
68  2020      8               27.4
69  2020      9               26.1
70  2020     10               25.6
71  2020     11               25.3
72  2020     12               27.3

[72 rows x 3 columns]

In [219]:
cheddar_production['In thousand tonnes'] = cheddar_production['In thousand tonnes'].astype(float)

In [220]:
cheddar_production['In thousand tonnes'].dtype

dtype('float64')

In [221]:
# And finally multiplying the values and creating in column 'In Kilogrammes'
cheddar_production['In Kilogrammes'] = cheddar_production['In thousand tonnes'] * 1000000
cheddar_production

Year  Month  In thousand tonnes  In Kilogrammes
1   2015      1                24.2       2.420e+07
2   2015      2                21.3       2.130e+07
3   2015      3                26.2       2.620e+07
4   2015      4                29.3       2.930e+07
5   2015      5                31.0       3.100e+07
..   ...    ...                 ...             ...
68  2020      8                27.4       2.740e+07
69  2020      9                26.1       2.610e+07
70  2020     10                25.6       2.560e+07
71  2020     11                25.3       2.530e+07
72  2020     12                27.3       2.730e+07

[72 rows x 4 columns]

In [222]:
cheddar_production.drop('In thousand tonnes', axis=1, inplace=True)

## Conversions

Let's take a look at our final df's of milk prices and cheddar production. 

In [223]:
milk_prices_FILTERED

Year  Month  Price in Pounds per Litre
26    2019     10                      0.294
451   2020      9                      0.290
481   2020      8                      0.281
506   2020      7                      0.277
515   2020      6                      0.271
...    ...    ...                        ...
1453  2019      1                      0.296
1458  2018     11                      0.316
1461  2017     10                      0.318
1490  2018     10                      0.315
1554  2019     11                      0.299

[71 rows x 3 columns]

In [224]:
cheddar_production

Year  Month  In Kilogrammes
1   2015      1       2.420e+07
2   2015      2       2.130e+07
3   2015      3       2.620e+07
4   2015      4       2.930e+07
5   2015      5       3.100e+07
..   ...    ...             ...
68  2020      8       2.740e+07
69  2020      9       2.610e+07
70  2020     10       2.560e+07
71  2020     11       2.530e+07
72  2020     12       2.730e+07

[72 rows x 3 columns]

In [225]:
conversion_factors

Product Conversion factor (litres/kg)
0             Butter (from Cream)                          2.04
1                         Cheddar                           9.5
2    Other Long Life Territorials                           9.2
3         Short life Territorials                           8.1
4                       Blue vein                           9.1
5                       Mozarella                           9.8
6   Cottage cheese /Fromage frais                           3.6
7                     Soft cheese                           8.0
8            Non-specified cheese                           9.7
9                  Condensed Milk                           2.7
10  Milk Powder (from Whole milk)                           8.5
11   Milk Powder (from Skim milk)                          10.1
12       Milk Powder (from Cream)                          12,3
13             Cream (from cream)                           1.0
14                        Yoghurt                           1,1

According to our conversion table we need 9.5 litres of milk to produce 1 kg of Cheddar. 

There are multiple entries from the same years. Different amount of cheddar produced in different months and the cost of milk varies in the months of the year. 

Hmmm. It seems like I've dropped the month, but I need it now... But no worries. I can take a quick step back and bring back the month in both tables, to correctly do the calculations.

> We need to calculate cost of the milk in the cheddar production each month based on the price in each month. 
    > But to be honest... We are not certain if the milk were bought the same month, the cheese was produced. But it is probably what is expcted from us in this task, so we'll make the calculation based on month. 
    
---

Back with month in both columns. 

---

Time for calculations. 

### Merging tables

In [226]:
tables_for_calculations = pd.merge(cheddar_production, milk_prices_FILTERED)

In [227]:
tables_for_calculations

Year  Month  In Kilogrammes  Price in Pounds per Litre
0   2015      1       2.420e+07                      0.265
1   2015      2       2.130e+07                      0.261
2   2015      3       2.620e+07                      0.251
3   2015      4       2.930e+07                      0.247
4   2015      5       3.100e+07                      0.241
..   ...    ...             ...                        ...
66  2020      7       2.890e+07                      0.277
67  2020      8       2.740e+07                      0.281
68  2020      9       2.610e+07                      0.290
69  2020     10       2.560e+07                      0.300
70  2020     11       2.530e+07                      0.306

[71 rows x 4 columns]

It is worth noting that the number of rows is not the same in production of cheddar table and milk price table (72/71). It is posiible that this will cause some trouble. But we'll see how it will work out and then apply some changes, look for bugs. 

--- 

It will also be appriopriate to make some changes in names of the columns to make it more clear. 

In [228]:
tables_for_calculations['Cheddar_In_Kg'] = tables_for_calculations['In Kilogrammes']
tables_for_calculations['Milk_Price_Per_Litre'] = tables_for_calculations['Price in Pounds per Litre']
tables_for_calculations['Date'] = tables_for_calculations['Year']
tables_for_calculations

Year  Month  In Kilogrammes  Price in Pounds per Litre  Cheddar_In_Kg  Milk_Price_Per_Litre  Date
0   2015      1       2.420e+07                      0.265      2.420e+07                 0.265  2015
1   2015      2       2.130e+07                      0.261      2.130e+07                 0.261  2015
2   2015      3       2.620e+07                      0.251      2.620e+07                 0.251  2015
3   2015      4       2.930e+07                      0.247      2.930e+07                 0.247  2015
4   2015      5       3.100e+07                      0.241      3.100e+07                 0.241  2015
..   ...    ...             ...                        ...            ...                   ...   ...
66  2020      7       2.890e+07                      0.277      2.890e+07                 0.277  2020
67  2020      8       2.740e+07                      0.281      2.740e+07                 0.281  2020
68  2020      9       2.610e+07                      0.290      2.610e+07                 0.290  2020
69  2020     10       2.560e+07                      0.300      2.560e+07                 0.300  2020
70  2020     11       2.530e+07                      0.306      2.530e+07                 0.306  2020

[71 rows x 7 columns]

In [229]:
# Tidyig up the table
tables_for_calculations.drop(['In Kilogrammes', 'Price in Pounds per Litre', 'Year'], axis=1, inplace=True)
tables_for_calculations

Month  Cheddar_In_Kg  Milk_Price_Per_Litre  Date
0       1      2.420e+07                 0.265  2015
1       2      2.130e+07                 0.261  2015
2       3      2.620e+07                 0.251  2015
3       4      2.930e+07                 0.247  2015
4       5      3.100e+07                 0.241  2015
..    ...            ...                   ...   ...
66      7      2.890e+07                 0.277  2020
67      8      2.740e+07                 0.281  2020
68      9      2.610e+07                 0.290  2020
69     10      2.560e+07                 0.300  2020
70     11      2.530e+07                 0.306  2020

[71 rows x 4 columns]

9.5 litres of milk to produce 1 kg of cheddar. 

So we need to multiply the amount of the kilograms of the cheddar produced by price of milk and the amount of milk needed for production of 1 kg. 

In [230]:
tables_for_calculations['cost'] = tables_for_calculations['Cheddar_In_Kg'] * tables_for_calculations['Milk_Price_Per_Litre'] * 9.5
tables_for_calculations

Month  Cheddar_In_Kg  Milk_Price_Per_Litre  Date       cost
0       1      2.420e+07                 0.265  2015  6.083e+07
1       2      2.130e+07                 0.261  2015  5.273e+07
2       3      2.620e+07                 0.251  2015  6.235e+07
3       4      2.930e+07                 0.247  2015  6.872e+07
4       5      3.100e+07                 0.241  2015  7.109e+07
..    ...            ...                   ...   ...        ...
66      7      2.890e+07                 0.277  2020  7.613e+07
67      8      2.740e+07                 0.281  2020  7.301e+07
68      9      2.610e+07                 0.290  2020  7.193e+07
69     10      2.560e+07                 0.300  2020  7.298e+07
70     11      2.530e+07                 0.306  2020  7.367e+07

[71 rows x 5 columns]

In [231]:
annual_cost = pd.DataFrame()

In [232]:
tables_for_calculations.groupby('Date').sum()

Month  Cheddar_In_Kg  Milk_Price_Per_Litre       cost
Date                                                       
2015     78      3.024e+08                 2.938  7.027e+08
2016     78      3.121e+08                 2.716  6.666e+08
2017     78      3.254e+08                 3.451  8.880e+08
2018     78      3.326e+08                 3.520  9.226e+08
2019     78      3.270e+08                 3.468  8.967e+08
2020     66      3.109e+08                 3.128  8.369e+08

In [233]:
annual_cost = tables_for_calculations.groupby('Date').sum()

In [234]:
annual_cost

Month  Cheddar_In_Kg  Milk_Price_Per_Litre       cost
Date                                                       
2015     78      3.024e+08                 2.938  7.027e+08
2016     78      3.121e+08                 2.716  6.666e+08
2017     78      3.254e+08                 3.451  8.880e+08
2018     78      3.326e+08                 3.520  9.226e+08
2019     78      3.270e+08                 3.468  8.967e+08
2020     66      3.109e+08                 3.128  8.369e+08

In [235]:
annual_cost.drop(['Month', 'Cheddar_In_Kg', 'Milk_Price_Per_Litre'], axis=1, inplace=True)

In [236]:
annual_cost['Cost'] = annual_cost['cost']
annual_cost.drop('cost', inplace=True, axis=1)
annual_cost

Cost
Date           
2015  7.027e+08
2016  6.666e+08
2017  8.880e+08
2018  9.226e+08
2019  8.967e+08
2020  8.369e+08